synapse/config/modules.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from synapse.config._base import Config, ConfigError
+from synapse.util.module_loader import load_module
+
+
+class ModulesConfig(Config):
+    section = "modules"
+
+    def read_config(self, config: dict, **kwargs):
+        self.loaded_modules = []
+
+        configured_modules = config.get("modules", [])
+        for i, module in enumerate(configured_modules):
+            config_path = ("modules", "<item %i>" % i)
+            if not isinstance(module, dict):
+                raise ConfigError("expected a mapping", config_path)
+
+            self.loaded_modules.append(load_module(module, config_path))
+
+    def generate_config_section(self, **kwargs):
+        return (
+            """
+            ## Modules ##
+            
+            # Server admins can expand Synapse's functionalities by using external modules
+            # to complement certain operations.
+            #
+            # See https://github.com/matrix-org/synapse/tree/master/docs/modules.md for
+            # more documentation on how to configure or create custom modules for Synapse.
+            #
+            modules:
+                # - module: my_super_module.MySuperClass
+                #   config:
+                #       do_thing: true
+                # - module: my_other_super_module.SomeClass
+                #   config: {}
```

Comment:
I've never really liked this `module`, `config` structure. I wonder if we should take the opportunity to reshuffle it. For example:

```yaml
modules:
    my_super_module.MySuperClass:
        do_thing: true
    my_other_super_module.SomeClass: {}
```

I'm not sure I love that either, though. Alternative suggestions welcome.

Reply:
(there must be other systems that have configurable modules that we can draw inspiration from?)

Reply:
Django typically puts everything (core and add-on settings) into the same settings namespace, for what that's worth. For example, see the Installation and Configuration sections for the Allauth module https://django-allauth.readthedocs.io/en/latest/installation.html

Reply:
I'm unsure if it matters, but it is probably worth considering whether we care about the order that modules run in, e.g. I'm unsure if the example above is ordered:

```yaml
modules:
    my_super_module.MySuperClass:
        do_thing: true
    my_other_super_module.SomeClass: {}
```

(Implementation-wise it is probably fine, but I don't know what YAML says about mappings.)

Reply:
Looks like the actual PR uses an array of objects (hence the `- ` prefixes), while Rich's snippet above is just a mapping, and as you pointed out, lord only knows what ordering we'll get out of that. :)

I'd argue in favor of the ordered array approach, which sidesteps this issue and potentially allows for multiple instances of the same module which vary only in their configuration.

Reply:
Thanks all, I'll keep it as is then.